<a href="https://colab.research.google.com/github/Daya-1-sgr/IPL_analysis/blob/main/IPL_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [2]:
df=pd.read_csv('/content/drive/MyDrive/IPL_analysis/deliveries.csv')
df.head()

match_id  inning           batting_team                 bowling_team  over  \
0    335982       1  Kolkata Knight Riders  Royal Challengers Bangalore     0   
1    335982       1  Kolkata Knight Riders  Royal Challengers Bangalore     0   
2    335982       1  Kolkata Knight Riders  Royal Challengers Bangalore     0   
3    335982       1  Kolkata Knight Riders  Royal Challengers Bangalore     0   
4    335982       1  Kolkata Knight Riders  Royal Challengers Bangalore     0   

   ball       batter   bowler  non_striker  batsman_runs  extra_runs  \
0     1   SC Ganguly  P Kumar  BB McCullum             0           1   
1     2  BB McCullum  P Kumar   SC Ganguly             0           0   
2     3  BB McCullum  P Kumar   SC Ganguly             0           1   
3     4  BB McCullum  P Kumar   SC Ganguly             0           0   
4     5  BB McCullum  P Kumar   SC Ganguly             0           0   

   total_runs extras_type  is_wicket player_dismissed dismissal_kind fielder  
0           1     legbyes          0              NaN            NaN     NaN  
1           0         NaN          0              NaN            NaN     NaN  
2           1       wides          0              NaN            NaN     NaN  
3           0         NaN          0              NaN            NaN     NaN  
4           0         NaN          0              NaN            NaN     NaN

In [3]:
df.sample(10)

match_id  inning                 batting_team  \
166985   1175365       1        Kolkata Knight Riders   
73989     548372       1  Royal Challengers Bangalore   
40030     419159       1              Kings XI Punjab   
37170     419147       1          Chennai Super Kings   
176128   1178422       2               Mumbai Indians   
203223   1254098       1        Kolkata Knight Riders   
19431     392208       1               Mumbai Indians   
121666    829817       2          Chennai Super Kings   
91304     598060       1          Sunrisers Hyderabad   
120480    829805       2        Kolkata Knight Riders   

                       bowling_team  over  ball         batter  \
166985               Delhi Capitals     8     6     KD Karthik   
73989              Delhi Daredevils     9     3        V Kohli   
40030           Chennai Super Kings    16     5      IK Pathan   
37170               Deccan Chargers    15     4       SK Raina   
176128        Kolkata Knight Riders    18     6        BB Sran   
203223          Chennai Super Kings     8     4     EJG Morgan   
19431   Royal Challengers Bangalore    12     2  ST Jayasuriya   
121666               Mumbai Indians     2     5   F du Plessis   
91304                Mumbai Indians     4     5      GH Vihari   
120480               Mumbai Indians     4     1      G Gambhir   

                 bowler   non_striker  batsman_runs  extra_runs  total_runs  \
166985         A Mishra  Shubman Gill             4           0           4   
73989   Y Venugopal Rao      CH Gayle             4           0           4   
40030      DE Bollinger      SE Marsh             2           0           2   
37170     Harmeet Singh   S Badrinath             1           0           1   
176128       AD Russell     KH Pandya             0           0           0   
203223        RA Jadeja   RA Tripathi             2           0           2   
19431          A Kumble      DJ Bravo             0           0           0   
121666       SL Malinga    MEK Hussey             1           0           1   
91304        MG Johnson      S Dhawan             0           0           0   
120480   MJ McClenaghan    RV Uthappa             1           0           1   

       extras_type  is_wicket player_dismissed dismissal_kind fielder  
166985         NaN          0              NaN            NaN     NaN  
73989          NaN          0              NaN            NaN     NaN  
40030          NaN          0              NaN            NaN     NaN  
37170          NaN          0              NaN            NaN     NaN  
176128         NaN          0              NaN            NaN     NaN  
203223         NaN          0              NaN            NaN     NaN  
19431          NaN          0              NaN            NaN     NaN  
121666         NaN          0              NaN            NaN     NaN  
91304          NaN          0              NaN            NaN     NaN  
120480         NaN          0              NaN            NaN     NaN

In [4]:
df.shape

(260920, 17)

In [5]:
df.columns

Index(['match_id', 'inning', 'batting_team', 'bowling_team', 'over', 'ball',
       'batter', 'bowler', 'non_striker', 'batsman_runs', 'extra_runs',
       'total_runs', 'extras_type', 'is_wicket', 'player_dismissed',
       'dismissal_kind', 'fielder'],
      dtype='object')

In [6]:
df['inning'].value_counts()

inning
1    135018
2    125741
3        77
4        72
5         8
6         4
Name: count, dtype: int64

Innings in a typical cricket match is 2 but the innings 3,4,5,6 requires further analysis

In [7]:
df[df['inning']==5].head()

match_id  inning    batting_team     bowling_team  over  ball  \
185372   1216517       5  Mumbai Indians  Kings XI Punjab     0     1   
185373   1216517       5  Mumbai Indians  Kings XI Punjab     0     2   
185374   1216517       5  Mumbai Indians  Kings XI Punjab     0     3   
185375   1216517       5  Mumbai Indians  Kings XI Punjab     0     4   
185376   1216517       5  Mumbai Indians  Kings XI Punjab     0     5   

            batter     bowler non_striker  batsman_runs  extra_runs  \
185372  KA Pollard  CJ Jordan   HH Pandya             1           0   
185373   HH Pandya  CJ Jordan  KA Pollard             0           1   
185374   HH Pandya  CJ Jordan  KA Pollard             1           0   
185375  KA Pollard  CJ Jordan   HH Pandya             4           0   
185376  KA Pollard  CJ Jordan   HH Pandya             0           1   

        total_runs extras_type  is_wicket player_dismissed dismissal_kind  \
185372           1         NaN          0              NaN            NaN   
185373           1       wides          0              NaN            NaN   
185374           1         NaN          0              NaN            NaN   
185375           4         NaN          0              NaN            NaN   
185376           1       wides          0              NaN            NaN   

       fielder  
185372     NaN  
185373     NaN  
185374     NaN  
185375     NaN  
185376     NaN

Searching for the match id 1216517 online we see that this MI vs KXIP match is a match with 2 super over match for only once in history of IPL

So,this clarifies the innings 3,4,5,6 in the data

How many SuperOvers has happened in 17 Seasons?

In [8]:
df[df['inning']==4]['match_id'].value_counts()

match_id
1082625    8
598004     6
598017     6
729315     6
1175365    6
1216517    6
1216547    6
1254077    6
392190     4
419121     4
829741     4
1216512    4
1178426    3
1216493    3
Name: count, dtype: int64

This also incudes the match with 2 super overs in a single match (MI vs KXIP Match no 36,2020)

In [9]:
df['over'].value_counts()

over
0     13906
1     13773
2     13597
3     13575
4     13560
5     13494
6     13452
7     13430
8     13396
9     13354
10    13289
11    13261
12    13222
13    13124
14    13024
15    12879
16    12685
17    12318
18    11583
19     9998
Name: count, dtype: int64

The overs feature seems to be correct

In [10]:
df['ball'].value_counts()

ball
1     42210
2     42106
3     42002
4     41891
5     41752
6     41619
7      7776
8      1307
9       225
10       30
11        2
Name: count, dtype: int64

On over has 6 legal deliveries ,So the remaining deliveries must be extras such as wide or noball etc

Which bowler has bowled most deliveries in an over?

In [11]:
df[df['ball']==11]['bowler'].value_counts()

bowler
Mohammed Siraj    1
TU Deshpande      1
Name: count, dtype: int64

In [12]:
df_first_over_data=df[df['over']==0]
df_first_over_data.tail(10)

match_id  inning                 batting_team          bowling_team  \
235263   1359513       1        Kolkata Knight Riders        Gujarat Titans   
217360   1304085       2  Royal Challengers Bangalore      Rajasthan Royals   
244563   1422122       1             Rajasthan Royals  Lucknow Super Giants   
259919   1426307       2          Sunrisers Hyderabad          Punjab Kings   
110940    829721       2             Rajasthan Royals        Mumbai Indians   
183300   1216509       2               Delhi Capitals   Chennai Super Kings   
162170   1136610       2              Kings XI Punjab        Mumbai Indians   
28710     419112       2  Royal Challengers Bangalore       Kings XI Punjab   
170697   1178400       2          Chennai Super Kings      Rajasthan Royals   
42432     501201       2               Mumbai Indians      Delhi Daredevils   

        over  ball        batter          bowler         non_striker  \
235263     0     3  N Jagadeesan  Mohammed Shami  Rahmanullah Gurbaz   
217360     0     5       V Kohli        TA Boult        F du Plessis   
244563     0     2   YBK Jaiswal     Mohsin Khan          JC Buttler   
259919     0     6   RA Tripathi  Arshdeep Singh     Abhishek Sharma   
110940     0     5     SV Samson   R Vinay Kumar           AM Rahane   
183300     0     4     AM Rahane       DL Chahar            S Dhawan   
162170     0     5      KL Rahul  MJ McClenaghan            CH Gayle   
28710      0     3     MK Pandey       IK Pathan           JH Kallis   
170697     0     4     SR Watson     DS Kulkarni        F du Plessis   
42432      0     4     DJ Jacobs        AB Dinda        SR Tendulkar   

        batsman_runs  extra_runs  total_runs extras_type  is_wicket  \
235263             0           0           0         NaN          0   
217360             4           0           4         NaN          0   
244563             1           0           1         NaN          0   
259919             0           1           1     legbyes          0   
110940             0           0           0         NaN          0   
183300             0           0           0         NaN          0   
162170             0           0           0         NaN          0   
28710              1           0           1         NaN          0   
170697             0           0           0         NaN          1   
42432              1           0           1         NaN          0   

       player_dismissed dismissal_kind fielder  
235263              NaN            NaN     NaN  
217360              NaN            NaN     NaN  
244563              NaN            NaN     NaN  
259919              NaN            NaN     NaN  
110940              NaN            NaN     NaN  
183300              NaN            NaN     NaN  
162170              NaN            NaN     NaN  
28710               NaN            NaN     NaN  
170697        SR Watson         bowled     NaN  
42432               NaN            NaN     NaN

In [18]:
batting_stats=df_first_over_data.groupby('batter').agg({'ball':'count','batsman_runs':'sum','is_wicket':'sum'}).reset_index()
batting_stats.head()

batter  ball  batsman_runs  is_wicket
0         A Chopra     8             2          0
1         A Mukund     1             1          0
2    A Raghuvanshi     5             7          0
3  AA Jhunjhunwala     1             0          0
4      AA Kulkarni     4             5          1

In [22]:
batting_stats['strike_rate']=(batting_stats['batsman_runs']/batting_stats['ball'])*100
batting_stats['batting_average']=batting_stats['batsman_runs']/batting_stats['is_wicket']
batting_stats[batting_stats['batsman_runs']>=50].sort_values(by='strike_rate',ascending=False).head()

batter  ball  batsman_runs  is_wicket  strike_rate  batting_average
251    YK Pathan    20            50          1   250.000000        50.000000
250  YBK Jaiswal   185           254          7   137.297297        36.285714
233      TM Head    48            64          3   133.333333        21.333333
92   JM Bairstow    85           110          1   129.411765       110.000000
166      PD Salt    62            80          5   129.032258        16.000000

In [17]:
df[df['batter']=='L Wood'].head(20)

match_id  inning    batting_team    bowling_team  over  ball  batter  \
254318   1426281       2  Mumbai Indians  Delhi Capitals    19     2  L Wood   
254319   1426281       2  Mumbai Indians  Delhi Capitals    19     3  L Wood   
254320   1426281       2  Mumbai Indians  Delhi Capitals    19     4  L Wood   

              bowler non_striker  batsman_runs  extra_runs  total_runs  \
254318  Mukesh Kumar   PP Chawla             2           0           2   
254319  Mukesh Kumar   PP Chawla             6           0           6   
254320  Mukesh Kumar   PP Chawla             1           0           1   

       extras_type  is_wicket player_dismissed dismissal_kind fielder  
254318         NaN          0              NaN            NaN     NaN  
254319         NaN          0              NaN            NaN     NaN  
254320         NaN          0              NaN            NaN     NaN

In [23]:
matches=pd.read_csv('/content/drive/MyDrive/IPL_analysis/matches.csv')
matches.head()

id   season        city        date match_type player_of_match  \
0  335982  2007/08   Bangalore  2008-04-18     League     BB McCullum   
1  335983  2007/08  Chandigarh  2008-04-19     League      MEK Hussey   
2  335984  2007/08       Delhi  2008-04-19     League     MF Maharoof   
3  335985  2007/08      Mumbai  2008-04-20     League      MV Boucher   
4  335986  2007/08     Kolkata  2008-04-20     League       DJ Hussey   

                                        venue                        team1  \
0                       M Chinnaswamy Stadium  Royal Challengers Bangalore   
1  Punjab Cricket Association Stadium, Mohali              Kings XI Punjab   
2                            Feroz Shah Kotla             Delhi Daredevils   
3                            Wankhede Stadium               Mumbai Indians   
4                                Eden Gardens        Kolkata Knight Riders   

                         team2                  toss_winner toss_decision  \
0        Kolkata Knight Riders  Royal Challengers Bangalore         field   
1          Chennai Super Kings          Chennai Super Kings           bat   
2             Rajasthan Royals             Rajasthan Royals           bat   
3  Royal Challengers Bangalore               Mumbai Indians           bat   
4              Deccan Chargers              Deccan Chargers           bat   

                        winner   result  result_margin  target_runs  \
0        Kolkata Knight Riders     runs          140.0        223.0   
1          Chennai Super Kings     runs           33.0        241.0   
2             Delhi Daredevils  wickets            9.0        130.0   
3  Royal Challengers Bangalore  wickets            5.0        166.0   
4        Kolkata Knight Riders  wickets            5.0        111.0   

   target_overs super_over method    umpire1         umpire2  
0          20.0          N    NaN  Asad Rauf     RE Koertzen  
1          20.0          N    NaN  MR Benson      SL Shastri  
2          20.0          N    NaN  Aleem Dar  GA Pratapkumar  
3          20.0          N    NaN   SJ Davis       DJ Harper  
4          20.0          N    NaN  BF Bowden     K Hariharan

In [25]:
most_potm=matches.groupby('player_of_match').agg({'id':'count'}).reset_index().sort_values(by='id',ascending=False)
most_potm.head(10)

player_of_match  id
11   AB de Villiers  25
43         CH Gayle  22
214       RG Sharma  19
53        DA Warner  18
276         V Kohli  18
165        MS Dhoni  17
252       SR Watson  16
286       YK Pathan  16
209       RA Jadeja  16
16       AD Russell  15

In [26]:
finisher_stats=df[df['over']>13]
finisher_stats.head()

match_id  inning           batting_team                 bowling_team  \
87    335982       1  Kolkata Knight Riders  Royal Challengers Bangalore   
88    335982       1  Kolkata Knight Riders  Royal Challengers Bangalore   
89    335982       1  Kolkata Knight Riders  Royal Challengers Bangalore   
90    335982       1  Kolkata Knight Riders  Royal Challengers Bangalore   
91    335982       1  Kolkata Knight Riders  Royal Challengers Bangalore   

    over  ball       batter    bowler  non_striker  batsman_runs  extra_runs  \
87    14     1    DJ Hussey  CL White  BB McCullum             4           0   
88    14     2    DJ Hussey  CL White  BB McCullum             1           0   
89    14     3  BB McCullum  CL White    DJ Hussey             6           0   
90    14     4  BB McCullum  CL White    DJ Hussey             4           0   
91    14     5  BB McCullum  CL White    DJ Hussey             0           2   

    total_runs extras_type  is_wicket player_dismissed dismissal_kind fielder  
87           4         NaN          0              NaN            NaN     NaN  
88           1         NaN          0              NaN            NaN     NaN  
89           6         NaN          0              NaN            NaN     NaN  
90           4         NaN          0              NaN            NaN     NaN  
91           2       wides          0              NaN            NaN     NaN

In [28]:
finishing_stats=finisher_stats.groupby('batter').agg({'ball':'count','batsman_runs':'sum','is_wicket':'sum'}).reset_index()
finishing_stats.head()

batter  ball  batsman_runs  is_wicket
0  A Ashish Reddy   164           243         14
1        A Badoni   292           439         16
2      A Chandila     7             4          1
3        A Chopra     3             2          1
4     A Choudhary    20            25          2

In [30]:
finishing_stats['strike_rate']=(finisher_stats['batsman_runs']/finisher_stats['ball'])*100
finishing_stats['batting_average']=finisher_stats['batsman_runs']/finisher_stats['is_wicket']
finishing_stats[finishing_stats['batsman_runs']>=50].sort_values(by='strike_rate',ascending=False).head()

batter  ball  batsman_runs  is_wicket  strike_rate  \
559         TH David   355           604         20        600.0   
343          MS Gony    64            94          8        600.0   
112        CH Morris   334           524         21        600.0   
87   B Sai Sudharsan   128           221         11        400.0   
344       MV Boucher   188           290         10        300.0   

     batting_average  
559              inf  
343              inf  
112              inf  
87               inf  
344              inf